### DISCUSSION TITLE
Langium example

### DISCUSSION THREAD

**Discussion by hallvard** [eclipse-langium/langium#149] [Show and tell]
Since I'm asking questions, I thought I'd say some words about the context. I'm porting an Xtext-based DSL for defining quizzes to Langium, to test it out. The DSL has a generator (not yet ported) for creating a web page that presents the questions and allow you to answer and check the result.

So far the DSL extension includes the LSP part, a command for running the generator and opening a preview pane on the generated HTML.

Here's the link: https://github.com/hallvard/langium-quiz

**Comment by spoenemann** [comment]
Thanks for sharing, @hallvard! I'd like to give a general hint on how we envision generators in Langium. Xtext included generators in the incremental build because that's how it was generally done in the Eclipse IDE. The tool landscape used in VS Code is different: tools offer a CLI with a `--watch` mode to incrementally rebuild parts of the project. This gives users far more control over what is generated and when. You can integrate such a generator CLI in the VS Code UI with [tasks](https://code.visualstudio.com/docs/editor/tasks).

We are going to provide examples for this (#131).

**Reply by hallvard** [reply]
I understand, the generator currently can be triggered manually by a command or automatically when the DSL file is changed. But the generator isn't a separate tool, it's part of the same extension. What's important is easy access to the AST. I hope I'm doing that part right. I'm totally new to typescript, so any feedback will help me learn!

**Reply by msujew** [reply]
Of course you need access to the AST in the CLI that @spoenemann proposes as well. You would ideally create a *library* part of your project that contains the generated langium code (AST, parser, etc.) and a another *vscode* and *cli* project that are all linked together through [npm workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces). Our Langium monorepo does exactly that.

**Reply by msujew** [reply]
Alternatively you could run everything within a single project, and add something like a `src/cli` directory in your vscode extension. That's what @spoenemann proposes in #157. Also that's the way it's implemented for the [arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/cli.ts).
